server/src/main/java/org/elasticsearch/cluster/routing/IndexRouting.java

```diff
@@ -251,109 +252,123 @@ public void process(IndexRequest indexRequest) {}
         public int indexShard(String id, @Nullable String routing, XContentType sourceType, BytesReference source) {
             assert Transports.assertNotTransportThread("parsing the _source can get slow");
             checkNoRouting(routing);
-            return hashToShardId(hashSource(sourceType, source));
+            return hashToShardId(hashSource(sourceType, source).buildHash(IndexRouting.ExtractFromSource::defaultOnEmpty));
         }
 
         public String createId(XContentType sourceType, BytesReference source, byte[] suffix) {
```

Comment:
This is the method that parses the source to extract the `_id`? Is it still used?

Reply:
Just in an assertion for now. I'm being a bit paranoid about it. I'd like to keep it for a little while, just because I think we might want to use it to rebuild the `_id` on the fly rather than loading it from a stored field.